asyncio_pipe/__init__.py

```diff
@@ -0,0 +1,172 @@
+
+            See also:
+                multiprocessing.Connection#send
+        """
+        await self._write_event.wait()
+        self._sync_connection.send(obj)
+        self._write_event.clear()
+
+    async def send_bytes(
+            self,
+            buf: bytes,
+            offset: int = 0,
+            size: Optional[int] = None) -> None:
+        """
+            Send the bytes data from a bytes-like object.
+
+            See also:
+                multiprocessing.Connection#send_bytes
+        """
+        await self._write_event.wait()
+        self._sync_connection.send_bytes(buf, offset, size)
+        self._write_event.clear()
+
+    async def poll(self, timeout: float = 0.0) -> bool:
+        """
+            Whether there is any input available to be read.
+
+            See also:
+                multiprocessing.Connection#poll
+        """
+        try:
+            await asyncio.wait_for(self._read_event.wait(), timeout=timeout)
+        except asyncio.TimeoutError:
+            return False
+        return True
+
+    async def recv(self) -> Any:
+        """
+            Receive a picklable object asynchronously.
+
+            See also:
+                multiprocessing.Connection#recv
+        """
+        await self._read_event.wait()
+        obj = self._sync_connection.recv()
+        self._read_event.clear()
+        return obj
+
+    async def recv_bytes(self, maxlength=None) -> bytes:
+        """
+            Receive bytes data as a bytes object.
+
+            See also:
+                multiprocessing.Connection#recv_bytes
+        """
+        await self._read_event.wait()
+        buf: bytes = self._sync_connection.recv_bytes(maxlength)
+        self._read_event.clear()
+        return buf
+
+    async def recv_bytes_into(self, buf, offset=0) -> int:
+        """
+            Receive bytes data into a writeable bytes-like object.
+            Return the number of bytes read.
+
+            See also:
+                multiprocessing.Connection#recv_bytes_into
+        """
+        await self._read_event.wait()
+        size: int = self._sync_connection.recv_bytes_into(buf, offset)
+        self._read_event.clear()
+        return size
+
+
+def AsyncPipe(
+        duplex: bool = True,
+        loop: Optional[asyncio.AbstractEventLoop] = None) -> Tuple[
+            AsyncConnection,
+            Connection]:
+    """
+        Returns pair of connection objects at either end of a pipe.
+        The first returned connection in asynchronous and thus can be used
+        without blocking an eventloop.
+
+        Parameters:
+            duplex (bool):
+                see multiprocessing.Pipe
+            loop (Optional[asyncio.AbstractEventLoop]):
+                Event loop to attach connection to.
+
+        Returns:
+            Tuple[AsyncConnection, Connection]:
+                Created connection pair.
+    """
+    c1, c2 = Pipe(duplex=duplex)
+    return AsyncConnection(c1, loop=loop), c2
```

Comment:
I was wondering actually when writing this naïve `Pipe` implementation if the case you describe would happen. So you think returning two `AsyncConnection` would be better ? This is a lot of case to cover (and especially when you are pickle the connection to send it to another process ?)

Reply:
I liked my initial approach of just wrapping a `multiprocessing.Connection` with `asyncio.Connection`. That way users can do it as necessary, on the correct event loop. 

I don't think we should re-implement the `Pipe` method. Returning two `AsyncConnection` objects doesn't seem useful because they would both be on the same event loop.